hw/ip/prim/doc/prim_keccak.md

```diff
@@ -0,0 +1,79 @@
+----
+title: "Primitive Component: Keccak permutation"
+----
+
+# Overview
+
+`prim_keccak` is a single round implementation of the permutation stage in [SHA3 algorithm][fibs-pub-202].
+Keccak primitive module assumes the Nr is less than or equal to 12 + 2L.
+It supports all combinations of the data width described in the [spec][fibs-pub-202].
+The implementation is not hardened against side-channel or fault injection attacks.
+It implements the Keccak_p function.
+
+[fibs-pub-202]: https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.202.pdf
+
+## Parameters
+
+Name  | Type | Description
+------|------|--------------
+Width | int  | state width. can be 25, 50, 100, 200, 400, 800, or 1600
+
+### Derived Parameters
+
+Name  | Type | Description
+------|------|--------------
+W     | int  | number of slices in state. `Width/25`
+L     | int  | log2 of `W`
+Round | int  | maximum allowed round value, assumed max of `12 + 2L`
+RndW  | int  | bit-width to represent Round. log2 of `Round + 1`
+
+## Signal Interfaces
+
+Signal | Type   | Description
+-------|--------|-------------
+rnd_i  | input  | current round number [0..23]
+s_i    | input  | state input
+s_o    | output | permutated state output
+
+# Theory of Operations
+
+```
+         |                                                          |
+rnd_i    |                                                          |
+---/---->| -----------------------------------------\               |
+ [RndW]  |                                          |               |
+         |                                          |               |
+s_i      |                                          V               | s_o
+===/====>| bit2s() -> chi(pi(rho(theta))) -> iota( ,rnd) -> s2bit() |==/==>
+ [Width] |            |-----------keccak_p--------------|           |[Width]
+         |
+```
+
+`prim_keccak` implements "Step Mappings" section in [SHA3 spec][fibs-pub-202].
+It is composed of five unique permutation functions, theta, rho, pi, chi, and iota.
+Also it has functions that converts bitstream of `Width` into `5x5xW` state and vice versa.
+
+Three constant parameters are defined inside the keccak primitive module.
+The rotate position described in phi function is hard-coded as below.
+The value is described in the SHA3 specification.
+
+```systemverilog
+localparam int PiRotate [5][5] = '{
+  //y  0    1    2    3    4     x
+  '{   0,   3,   1,   4,   2},// 0
+  '{   1,   4,   2,   0,   3},// 1
+  '{   2,   0,   3,   1,   4},// 2
+  '{   3,   1,   4,   2,   0},// 3
+  '{   4,   2,   0,   3,   1} // 4
+};
+```
+
+The shift amount in rho function is defined as `RhoOffset` parameter.
+The value is same as in the specification, but it is used after the modular of `W`.
```

Comment:
I'm not sure what "used after the modular of `W`" means? Having trouble
parsing that sentence.